### PR TITLE
Use firestore.

### PR DESCRIPTION
**Pull request by SurferJeffAtGoogle**

### CODE REVIEW

File: sessionstate-core/SessionState/Services/FirestoreDistributedCache.cs [hunk 1]
```diff
@@ -28,29 +28,117 @@
...
-    class DatastoreDistributedCacheOptions
+    class FirestoreDistributedCacheOptions
     {
         /// <summary>
         /// Your Google project id.
         /// </summary>
         public string ProjectId { get; set; }
+    }
+
+    // A class like this should be provided by the library.  I shouldn't have
```
Comment (SurferJeffAtGoogle) [inline sessionstate-core/SessionState/Services/FirestoreDistributedCache.cs:39]: A scoped Transaction class like this needs to be provided by the library.  Every C# developer wants a Transaction they can use in a "using statement."

File: sessionstate-core/SessionState/Services/FirestoreDistributedCache.cs [hunk 2]
```diff
@@ -28,29 +28,117 @@
...
-        private KeyFactory _sessionExpiresKeyFactory;
+        internal FirestoreClient _firestore;
+        internal ILogger _logger;
 
-        private ILogger _logger;
+        internal string _projectId;
+
+        public string SessionDatabaseName 
+        { 
+            get => $"projects/{_projectId}/databases/sessions";
```
Comment (SurferJeffAtGoogle) [inline sessionstate-core/SessionState/Services/FirestoreDistributedCache.cs:133]: I want some help from the library for generating this string.

File: sessionstate-core/SessionState/Services/FirestoreDistributedCache.cs [hunk 3]
```diff
@@ -95,149 +181,210 @@ private const string
             }
         }
 
-        Key[] ToEntityKeys(string name) => new Key[] {
-            _sessionKeyFactory.CreateKey(name), 
-            _sessionExpiresKeyFactory.CreateKey(name)
-        };
+        string ToSessionDocName(string key) =>
+            $"{SessionDatabaseName}/documents/{key}";
```
Comment (SurferJeffAtGoogle) [inline sessionstate-core/SessionState/Services/FirestoreDistributedCache.cs:185]: Maybe more help from the library for this string too.

File: sessionstate-core/SessionState/Services/FirestoreDistributedCache.cs [hunk 4]
```diff
@@ -95,149 +181,210 @@ private const string
...
+        string SweepLockDocName { 
+            get => $"${SessionSweepDatabaseName}/documents/sweeplock";            
+        }
 
         public byte[] Get(string key) 
         {
             _logger.LogDebug("Get({0})", key);
-            return BytesFromEntities(_datastore.Lookup(ToEntityKeys(key), 
-                ReadConsistency.Strong, _callSettings));
+            var doc = _firestore.GetDocument(new GetDocumentRequest() {
```
Comment (SurferJeffAtGoogle) [inline sessionstate-core/SessionState/Services/FirestoreDistributedCache.cs:194]: Too tedious.  I want:
```cs
var doc = _firestore.GetDocument(ToSessionDocName(key));

File: sessionstate-core/SessionState/Services/FirestoreDistributedCache.cs [hunk 5]
```diff
@@ -95,149 +181,210 @@ private const string
...
+                    Name = ToSessionDocName(key),
+                    Transaction = transaction.Transaction
+                });
+                if (UpdateExpiration(doc))
+                {
+                    transaction.Writes.Add(new Write() {
+                        Update = doc,
+                        // I really dislike update masks.  Is there an easy way
+                        // to say 'update everything'?
+                        UpdateMask = new DocumentMask() {
```
Comment (SurferJeffAtGoogle) [inline sessionstate-core/SessionState/Services/FirestoreDistributedCache.cs:225]: Is there an easy way to say "Update everything?"

File: sessionstate-core/SessionState/Services/FirestoreDistributedCache.cs [hunk 6]
```diff
@@ -95,149 +181,210 @@ private const string
...
+                    await transaction.CommitAsync();
+                }
+            }
+        }
+
+        // This is a very complicated way to do a write.
+        // And I suspect it's very inefficient.
+        async Task<WriteResponse> WriteAsync(Write write) 
+        {
+            var writeStream = _firestore.Write();
```
Comment (SurferJeffAtGoogle) [inline sessionstate-core/SessionState/Services/FirestoreDistributedCache.cs:260]: This is a really painful way to write.  And I suspect it's very inefficient.  Making in more efficient looks pretty complicated.  How would I?

File: sessionstate-core/SessionState/Services/FirestoreDistributedCache.cs [hunk 7]
```diff
@@ -95,149 +181,210 @@ private const string
...
+                expires = DateTime.UtcNow.Add(options.SlidingExpiration.Value);
             }
             else
             {
                 throw new ArgumentException("Required expiration option was not set.", "options");
             }
-            return new [] {session, sessionExpires};
+            // Converting between C# Timestamp times and Proto timestamp types
+            // is hideous.
+            doc.Fields[EXPIRATION] = new Value() {
```
Comment (SurferJeffAtGoogle) [inline sessionstate-core/SessionState/Services/FirestoreDistributedCache.cs:360]: Converting between proto time types and C# time types is soo painful.

File: sessionstate-core/SessionState/Services/FirestoreDistributedCache.cs [hunk 8]
```diff
@@ -257,46 +404,81 @@ bool UpdateExpiration(Entity sessionExpires)
...
                 }
+#if false                
                 try
                 {
                     _logger.LogInformation("Beginning sweep.");
                     // Find old sessions to clean up
                     var now = DateTime.UtcNow;
                     var grace_period = TimeSpan.FromMinutes(20);
+                    // Wow.  Running a query is a huge mess.
+                    _firestore.RunQuery(new RunQueryRequest() {
```
Comment (SurferJeffAtGoogle) [inline sessionstate-core/SessionState/Services/FirestoreDistributedCache.cs:462]: Trying to compose a query is really, really painful.

File: sessionstate-core/SessionState/Services/FirestoreDistributedCache.cs [hunk 9]
```diff
@@ -257,46 +404,81 @@ bool UpdateExpiration(Entity sessionExpires)
...
+                        StructuredQuery = new StructuredQuery() {
+                            Where = new StructuredQuery.Types.Filter() {
```
Comment (SurferJeffAtGoogle) [inline sessionstate-core/SessionState/Services/FirestoreDistributedCache.cs:464]: How do I limit my request to just sessions?

File: sessionstate-core/SessionState/Services/FirestoreDistributedCache.cs [hunk 10]
```diff
@@ -330,6 +512,7 @@ bool UpdateExpiration(Entity sessionExpires)
                 {
                     _logger.LogError(3, e, "Failed to query expired sessions.");
                 }
+#endif
```
Comment (SurferJeffAtGoogle) [inline sessionstate-core/SessionState/Services/FirestoreDistributedCache.cs:515]: I haven't found the right code to do a query yet.

**Review by SurferJeffAtGoogle — COMMENTED** [review]
This felt a lot like datastore code before datastore got hand-written wrappers.

The main pain points are:
1. There's no such thing as a simple write.  Writes are very complicated, and hard to make efficient.
2. Querying is very messy.
3. Converting time types is a hassle.
4. Need a Transaction class that working with C#'s using statement.
5. It would be nice to have a Document class wrapper that automatically composed and DocumentMask as I make changes to the Document instance.

**Comment by samtstern** [comment]
cc @jskeet

**Comment by jskeet** [comment]
I've only just seen this PR - it looks like it's using the low-level Firestore API instead of the Google.Cloud.Firestore abstraction. I think we want to be guiding users to the higher-level abstraction.

See https://googlecloudplatform.github.io/google-cloud-dotnet/docs/Google.Cloud.Firestore/userguide.html for some brief examples of using the high-level API.

(I'm at a conference this week, so don't have much time to dive into things, but we can set up a meeting for next Monday or Tuesday if it would be helpful.)

**Comment by SurferJeffAtGoogle** [comment]
OMG.  I completely missed that, because I only looked in the
Google.Cloud.Firestore.V1Beta1 namespace.  I didn't expect to find anything
in Google.Cloud.Firestore.  I'll try again with the higher level
abstraction.

**Pull request closed without merging**